File: online-judge/4th-semester/exam3/11402_AhoyPirates2.py
#Made by Jhoan Buitrago 12/06/2021 dd/mm/aaaa

#Upgraded correct answer but still slow because python
#C++ version works nice

from sys import stdin, stdout

MAXN = 1030000
tree = [None for _ in range(MAXN * 2)]
pend = ["$"] * (MAXN * 2)

def reset():
    global pend
    pend = ["$"] * (MAXN * 2)

def invValue():
    return (0, 0) 

def createValue(val):
    #(Buccaneer, Barbary)
    if val == '0':
        return (0, 1)
    else:
        return (1, 0)

def combine(p1, p2):
    return (p1[0] + p2[0], p1[1] + p2[1])

def push(v, v1, v2, l, r):
    select(v1, pend[v])
    #print(f"pend[{v1}] is {pend[v1]}")
    
    select(v2, pend[v])
    #print(f"pend[{v2}] is {pend[v2]}")
    #print("Push", tree[v])
    change(v1, pend[v])
    change(v2, pend[v])
    
    #print(f"tree[{v}] is {tree[v]}")
    pend[v] = "$"

def change(v, letter):
    ans = (tree[v][0], tree[v][1])
    if letter == "F":
        tree[v] = (ans[0] + ans[1], 0)
    elif letter == "E":
        tree[v] = (0, ans[0] + ans[1]) 
    elif letter == "I":
        tree[v] = (ans[1], ans[0]) 

def select(v, letter):
    if letter == "F" or letter == "E":
        pend[v] = letter
    elif letter == "I":
        if pend[v] == "I":
            pend[v] = "$"

        elif pend[v] == "F":
            pend[v] = "E"
       
        elif pend[v] == "E":
            pend[v] = "F"
        
        elif pend[v] == "$":
            pend[v] = "I"

    #elif letter == "$": do nothing

#build the segment tree
def build(a, v, l, r):
    if l == r: 
        tree[v] = createValue(a[l])
    else:
        m = l + ((r - l) >> 1)
        build(a, v + 1, l, m)
        build(a, v + 2 * (m - l + 1), m + 1, r)
        tree[v] = combine(tree[v + 1], tree[v + 2 * (m - l + 1)])

#query
def query(v, L, R, l, r):
    ans = None
    if l > r: 
        ans = invValue()
    elif l == L and r == R:
        # m = L + ((R - L) >> 1)
        # push(v, v + 1, v + 2 * (m - L + 1), L, R)
        ans = tree[v]
    else:
        m = L + ((R - L) >> 1)
        push(v, v + 1, v + 2 * (m - L + 1), L, R)
        ans = combine(query(v + 1, L, m, l, min(r, m)), 
                      query(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r))
        #print(f"combined {l} {r} = {ans}")
    return ans

def update(v, L, R, l, r, h : str):
    if l <= r:
        if l == L and r == R: 
            select(v, h)
            #change(v, pend[v])
            change(v, h)
        else:
            m = L + ((R - L) >> 1)
            push(v, v + 1, v + 2 * (m - L + 1), L, R)
            update(v + 1, L, m, l, min(r, m), h)
            update(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r, h)
            tree[v] = combine(tree[v + 1], tree[v + 2 * (m - L + 1)])
            #print(f"combined {l} {r} = {tree[v]}")
            

def main():
    cases = int(stdin.readline())
    for i in range(1, cases + 1):
        if i != 1:
            reset()
        string = ""
        mPairs = int(stdin.readline())
        for _ in range(mPairs):
            mult = int(stdin.readline())
            subString = stdin.readline().rstrip()
            string += (subString * mult)
        last = len(string) - 1
        #print(string)
        build(string, 0, 0, last)
        queries = int(stdin.readline())
        count = 1
        stdout.write(f"Case {i}:\n")
        for _ in range(queries):
            info = stdin.readline().split()
            op = info[0]
            left = int(info[1])
            right = int(info[2])
            #print(op, left, right)
            if op == "S":
                ans = query(0, 0, last, left, right)[0]
                stdout.write(f"Q{count}: {ans}\n")
                count += 1
            else:
                update(0, 0, last, left, right, op)
                #print("after update pend[0] is", pend[0])
            # for j in range(last * 2 + 3):
            #     print(tree[j], end=" ")
            # print()


main()

File: online-judge/4th-semester/exam3/11402_AhoyPirates2.cpp
//Made by Jhoan Buitrago 12/06/2021 dd/mm/aaaa

//Works just fine :)


#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

const int MAXN = 1030000;

vector<pair<int, int>> tree(MAXN * 2);
vector<char> pend(MAXN * 2, '$');

void reset()
{
    for(int i = 0; i < MAXN * 2; i++)
    {
        pend[i] = '$';
    }
}

pair<int, int> invValue()
{
    return make_pair(0, 0);
}

pair<int, int> createValue(int val)
{
    if(val == '0') 
        return make_pair(0, 1);
    else 
        return make_pair(1, 0);
}

pair<int, int> combine(pair<int, int> p1, pair<int, int> p2)
{
    return make_pair(p1.first + p2.first, p1.second + p2.second);
}

void selectLetter(int v, char letter)
{
    if (letter == 'F' || letter == 'E')
        pend[v] = letter;
    else if (letter == 'I')
    {
        if (pend[v] == 'I')
            pend[v] = '$';

        else if (pend[v] == 'F')
            pend[v] = 'E';
       
        else if (pend[v] == 'E')
            pend[v] = 'F';
        
        else if (pend[v] == '$')
            pend[v] = 'I';
    }
}

void change(int v, char letter)
{
    
    pair<int, int> ans = tree[v];
    if (letter == 'F')
    {
        ans.first += ans.second;
        ans.second = 0;
    }
    else if (letter == 'E')
    {
        ans.second += ans.first;
        ans.first = 0;
    }
    else if (letter == 'I')
    {
        int tmp = ans.first;
        ans.first = ans.second;
        ans.second = tmp;
    }
    tree[v] = ans;
}

void push(int v, int v1, int v2)
{
    selectLetter(v1, pend[v]);
    selectLetter(v2, pend[v]);
    change(v1, pend[v]);
    change(v2, pend[v]);
    
    pend[v] = '$';
}

void build(string &arr, int v, int l, int r)
{
    if(l == r)
    {
        tree[v] = createValue(arr[l]);
    }
    else
    {
        int mid = l + ((r - l) >> 1);
        build(arr, v + 1, l, mid);
        build(arr, v + 2 * (mid - l + 1), mid + 1, r);
        tree[v] = combine(tree[v + 1], tree[v + 2 * (mid - l + 1)]);
    }
}

pair<int, int> query(int v, int L, int R, int l, int r)
{
    pair<int, int> ans;
    if(l > r)
    {
        ans = invValue();
    }
    else if(l == L && r == R) 
    {
        ans = tree[v];
    }
    else
    {
        int m = L + ((R - L) >> 1);
        push(v, v + 1, v + 2 * (m - L + 1));
        ans = combine(query(v + 1, L, m, l, min(r, m)), 
                      query(v + 2 * (m - L + 1), m + 1, R, max(l, m + 1), r));

    }
    return ans;
}

void update(int v, int L, int R, int l, int r, char letter)
{
    if(l <= r)
    {
        if(l == L && r == R)
        {
            selectLetter(v, letter);
            change(v, letter);
            
        }
        else
        {
            int mid = L + ((R - L) >> 1);
            push(v, v + 1, v + 2 * (mid - L + 1));
            update(v + 1, L, mid, l, min(r, mid), letter);
            update(v + 2 * (mid - L + 1), mid + 1, R, max(l, mid + 1), r, letter);
            tree[v] = combine(tree[v + 1], tree[v + 2 * (mid - L + 1)]);
        }
    }
}


int main()
{
    std::ios_base::sync_with_stdio(false);
    int cases, mPairs, mult, last, queries;
    int left, right, ans;
    int countGod, count;
    count = 1;
    string op;
    char letter;
    cin >> cases;
    while(cases--)
    {
        if(count != -1)
        {
            reset();
        }
        string arr = "";
        cin >> mPairs;
        while(mPairs--)
        {
            cin >> mult;
            cin >> op;
            for(int i = 0; i < mult; i++)
            {
                arr += op;
            }
        }
        last = arr.size() - 1;
        build(arr, 0, 0, last);
        countGod = 1;
        cin >> queries;
        cout << "Case " << count++ << ":\n";
        while(queries--)
        {
            cin >> letter >> left >> right;
            if(letter == 'S')
            {
                ans = query(0, 0, last, left, right).first;
                cout << "Q" << countGod++ << ": " << ans << "\n";
            }
            else
            {
                update(0, 0, last, left, right, letter);
            }
        }
    }
    return 0;
}